man/queue/sp_queue_qinsert.3
.\"  Staple - A general-purpose data structure library in pure C89.
.\"  Copyright (C) 2021  Randoragon
.\"
.\"  This library is free software; you can redistribute it and/or
.\"  modify it under the terms of the GNU Lesser General Public
.\"  License as published by the Free Software Foundation;
.\"  version 2.1 of the License.
.\"
.\"  This library is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\"  Lesser General Public License for more details.
.\"
.\"  You should have received a copy of the GNU Lesser General Public
.\"  License along with this library; if not, write to the Free Software
.\"  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
.\"--------------------------------------------------------------------------------
.TH SP_QUEUE_QINSERT 3 DATE "libstaple-VERSION"
.SH NAME
sp_queue_qinsert,
sp_queue_qinsertc,
sp_queue_qinserts,
sp_queue_qinserti,
sp_queue_qinsertl,
sp_queue_qinsertsc,
sp_queue_qinsertuc,
sp_queue_qinsertus,
sp_queue_qinsertui,
sp_queue_qinsertul,
sp_queue_qinsertf,
sp_queue_qinsertd,
sp_queue_qinsertld,
sp_queue_qinsertll,
sp_queue_qinsertull,
sp_queue_qinsertu8,
sp_queue_qinsertu16,
sp_queue_qinsertu32,
sp_queue_qinsertu64,
sp_queue_qinserti8,
sp_queue_qinserti16,
sp_queue_qinserti32,
sp_queue_qinserti64,
sp_queue_qinsertstr,
sp_queue_qinsertstrn
\- quickly insert an element at any position of a queue
.SH SYNOPSIS
.ad l
#include <staple.h>
.sp
int
.BR sp_queue_qinsert "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
const void
.RI * elem )
.br
int
.BR sp_queue_qinsertc "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
char
.IR elem )
.br
int
.BR sp_queue_qinserts "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
short
.IR elem )
.br
int
.BR sp_queue_qinserti "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int
.IR elem )
.br
int
.BR sp_queue_qinsertl "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
long
.IR elem )
.br
int
.BR sp_queue_qinsertsc "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
signed char
.IR elem )
.br
int
.BR sp_queue_qinsertuc "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
unsigned char
.IR elem )
.br
int
.BR sp_queue_qinsertus "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
unsigned short
.IR elem )
.br
int
.BR sp_queue_qinsertui "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
unsigned int
.IR elem )
.br
int
.BR sp_queue_qinsertul "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
unsigned long
.IR elem )
.br
int
.BR sp_queue_qinsertf "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
float
.IR elem )
.br
int
.BR sp_queue_qinsertd "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
double
.IR elem )
.br
int
.BR sp_queue_qinsertld "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
long double
.IR elem )
.br
int
.BR sp_queue_qinsertll "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
long long
.IR elem )
.br
int
.BR sp_queue_qinsertull "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
unsigned long long
.IR elem )
.br
int
.BR sp_queue_qinsertu8 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
uint8_t
.IR elem )
.br
int
.BR sp_queue_qinsertu16 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
uint16_t
.IR elem )
.br
int
.BR sp_queue_qinsertu32 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
uint32_t
.IR elem )
.br
int
.BR sp_queue_qinsertu64 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
uint64_t
.IR elem )
.br
int
.BR sp_queue_qinserti8 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int8_t
.IR elem )
.br
int
.BR sp_queue_qinserti16 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int16_t
.IR elem )
.br
int
.BR sp_queue_qinserti32 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int32_t
.IR elem )
.br
int
.BR sp_queue_qinserti64 "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
int64_t
.IR elem )
.br
int
.BR sp_queue_qinsertstr "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
const char
.RI * elem )
.br
int
.BR sp_queue_qinsertstrn "(struct sp_queue"
.RI * queue ,
size_t
.IR idx ,
const char
.RI * elem ,
size_t
.IR len )
.sp
Link with \fI-lstaple\fP.
.ad
.SH DESCRIPTION
See
.BR sp_queue_insert (3).
.P
The functions in the
.BR sp_queue_qinsert ()
family operate in constant time, which is achievable by changing the order of
the elements upon insertion.
.SH CONFORMING TO
See
.BR libstaple (7).
.SH EXAMPLES
To explain how the
.BR sp_queue_insert ()
and
.BR sp_queue_qinsert ()
families differ, consider the following initial queue (from head to tail):
.IP
.B 1 2 3 4 _ _ _
.P
Now, let us investigate what happens when we try to insert the number
.B 5
at position
.IR 2 .
.P
The functions in the
.BR sp_queue_insert ()
family would move
.BR 3 " and " 4
backwards, and insert
.B 5
without altering the order:
.IP (1)
.B 1 2 _ 3 4 _ _
.br
\h'6n'^ ^
.IP (2)
.B 1 2 5 3 4 _ _
.br
\h'4n'^
.P
On the contrary, the functions in the
.BR sp_queue_qinsert ()
family achieve the goal faster by moving the element "in the way" (in this case
.BR 3 )
to the back of the queue:
.IP (1)
.B 1 2 _ 4 3 _ _
.br
\h'8n'^
.IP (2)
.B 1 2 5 4 3 _ _
.br
\h'4n'^
.P
The second approach guarantees constant time, but does not preserve the order of
the elements.
.SH SEE ALSO
.ad l
.BR libstaple (7),
.BR sp_queue (7),
.BR sp_queue_create (3),
.BR sp_queue_destroy (3),
.BR sp_queue_clear (3),
.BR sp_queue_push (3),
.BR sp_queue_peek (3),
.BR sp_queue_pop (3),
.BR sp_queue_insert (3),
.BR sp_queue_remove (3),
.BR sp_queue_qremove (3),
.BR sp_queue_get (3),
.BR sp_queue_set (3),
.BR sp_queue_eq (3),
.BR sp_queue_copy (3),
.BR sp_queue_map (3),
.BR sp_queue_print (3)


src/queue/sp_queue_qinsert.c
/*  Staple - A general-purpose data structure library in pure C89.
 *  Copyright (C) 2021  Randoragon
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation;
 *  version 2.1 of the License.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "../sp_queue.h"
#include "../internal.h"
#include "../sp_errcodes.h"
#include <string.h>

int sp_queue_qinsert(struct sp_queue *queue, size_t idx, const void *elem)
{
	char *p;
#ifdef STAPLE_DEBUG
	if (queue == NULL) {
		error(("queue is NULL"));
		return SP_EINVAL;
	}
	if (elem == NULL) {
		error(("elem is NULL"));
		return SP_EINVAL;
	}
	if (idx > queue->size) {
		error(("index out of range"));
		return SP_EINDEX;
	}
#endif
	if (sp_size_try_add(queue->size * queue->elem_size, queue->elem_size))
		return SP_ERANGE;
	if (sp_ringbuf_fit(&queue->data, queue->size, &queue->capacity, queue->elem_size, &queue->head, &queue->tail))
		return SP_ENOMEM;
	if (queue->size != 0)
		sp_ringbuf_incr(&queue->tail, queue->data, queue->capacity, queue->elem_size);
	p = sp_ringbuf_get(idx, queue->data, queue->capacity, queue->elem_size, queue->head);
	memcpy(queue->tail, p, queue->elem_size);
	memcpy(p, elem, queue->elem_size);
	++queue->size;
	return 0;
}
